### ISSUE TITLE
[cs] Issue with enumerating over maps with abstract keys

### ISSUE THREAD

**Issue by dogles** [HaxeFoundation/haxe#3637]
Repro (latest Haxe development branch):

```
abstract Abs(String) to String {
  inline public function new(s:String) this = s;
  public function raw() return this;
}


class TestBug
{
  public static function main() {
    var map = new Map<Abs, Int>();
    map[new Abs("a")] = 1;
    map[new Abs("b")] = 1;

    for (key in map.keys()) { // ERROR Void -> Iterator<String> should be Void -> Iterator<Abs>

      trace(key);
      trace(map[key]);
    }
  }
}
```

This fails to compile in C#, but succeeds in Javascript. I think this is a new issue arising from the classes made for the map iterators.

**Comment by dogles** [comment]
@waneck @nadako Can we get this in for 3.2 (if it hasn't been already fixed)? If the bug wasn't clear, this means abstracts can't be used as map keys in the C# target.

**Comment by nadako** [comment]
This still reproduces on latest commit.

**Comment by waneck** [comment]
@Simn any idea what goes on in here?

**Comment by Simn** [comment]
I'm pretty sure inlining is going on. Like the same as #3713.

**Comment by ncannasse** [comment]
@dogles we'll make sure to fix this for 3.2
